src/FffGcodeWriter.cpp

```diff
@@ -143,83 +147,107 @@ void FffGcodeWriter::writeGCode(SliceDataStorage& storage, TimeKeeper& time_keep
+        storage.spiralize_wall_outlines[layer_nr] = nullptr;
+        storage.spiralize_seam_vertex_indices[layer_nr] = 0;
 
-            while (a > M_PI)
+        // iterate through extruders until we find a mesh that has a part with insets
+        const std::vector<unsigned int> &extruder_order = extruder_order_per_layer[layer_nr];
+        for (unsigned int extruder_idx = 0; !done_this_layer && extruder_idx < extruder_order.size(); ++extruder_idx)
+        {
+            const unsigned int extruder_nr = extruder_order[extruder_idx];
+            // iterate through this extruder's meshes until we find a part with insets
+            const std::vector<unsigned int>& mesh_order = mesh_order_per_extruder[extruder_nr];
+            for (unsigned int mesh_idx : mesh_order)
             {
-                // the vertex was on the right of the vector so move the seam vertex on
-                seam_vertex_idx = (seam_vertex_idx + 1) % n_points;
-                a = LinearAlg2D::getAngleLeft(last_wall_seam_vertex_vector, last_wall_seam_vertex, wall[(seam_vertex_idx + 1) % n_points]);
-
-                if (seam_vertex_idx == first_seam_vertex_idx)
+                SliceMeshStorage& mesh = storage.meshes[mesh_idx];
+                // if this mesh has layer data for this layer process it
+                if (!done_this_layer && mesh.layers.size() > layer_nr)
                 {
-                    logWarning("WARNING: findLayerSeamsForSpiralize() failed to find a suitable seam vertex on layer %d\n", layer_nr);
-                    // this shouldn't happen!
-                    break;
+                    SliceLayer& layer = mesh.layers[layer_nr];
+                    // if the first part in the layer (if any) has insets, process it
+                    if (layer.parts.size() != 0 && layer.parts[0].insets.size() != 0)
+                    {
+                        // last_layer_nr will be < 0 until we have processed the first non-empty layer
+                        if (last_layer_nr < 0)
+                        {
+                            // If the user has specified a z-seam location, use the vertex closest to that location for the seam vertex
+                            // in the first layer that has a part with insets. This allows the user to alter the seam start location which
+                            // could be useful if the spiralization has a problem with a particular seam path.
+                            Point seam_pos(0, 0);
+                            if (mesh.getSettingAsZSeamType("z_seam_type") == EZSeamType::USER_SPECIFIED)
+                            {
+                                seam_pos = Point(mesh.getSettingInMicrons("z_seam_x"), mesh.getSettingInMicrons("z_seam_y"));
+                            }
+                            storage.spiralize_seam_vertex_indices[layer_nr] = PolygonUtils::findClosest(seam_pos, layer.parts[0].insets[0][0]).point_idx;
+                        }
+                        else
```

Comment:
This whole function is becoming quite large.
Could you please factor a piece of this code into a new function?
Perhaps something like `findSpiralizeLayerStart(...)`